pkg/core/tokens/validator.go

```diff
@@ -33,24 +38,34 @@ var _ Validator = &jwtTokenValidator{}
 
 func (j *jwtTokenValidator) ParseWithValidation(ctx context.Context, rawToken Token, claims Claims) error {
 	token, err := jwt.ParseWithClaims(rawToken, claims, func(token *jwt.Token) (interface{}, error) {
-		serialNumber, err := tokenSerialNumber(token, claims)
+		serialNumberRaw, exists := token.Header[KeyIDHeader]
+		if !exists {
+			if _, ok := claims.(KeyIDFallback); ok {
+				// KID wasn't supported in the past, so we use a marker interface to indicate which tokens were allowed
+				// This will be removed with https://github.com/kumahq/kuma/issues/5519
+				log.Info("[WARNING] Using token with KID header, you should rotate this token as it will not be valid in future versions of Kuma", "jti", claims.ID(), KeyIDHeader, 0)
```

Comment:
I'd consider logging claims. It's safe to log, it does not contain secret. It might be a bit hard to locate a token by ID.

Additionally, we don't now what type of token it is